Index Checker creates `@LessThan(doubleValue)`

Consider:
```java
package math;

import java.math.RoundingMode;

public class UnparsableBug {
    double roundIntermediate(double x, RoundingMode mode){
        if (x <= 0.0) {
            return x;
        } else {
            return (long) x + 1;
        }

    }
    public static void main(String args[]) {

    }
}
```
Warning:
```java
Error:(15, 22) java: [expression.unparsable.type.invalid] Expression in dependent type annotation invalid: [error for expression: 0.0; error: '0']
```

@NhatDinh Why is the output that you provided expected?

I get the following output:
```
bug.java:6: warning: [cast.unsafe] "@LessThan("0.0") double" may not be casted to the type "@LessThan("[error for expression: 0.0; error: '0']") long"
			return (long) x - 1;
			       ^
bug.java:6: error: [expression.unparsable.type.invalid] Expression in dependent type annotation invalid: [error for expression: 0.0; error: '0']
			return (long) x - 1;
			        ^
1 error
1 warning
```

If this is a problem with expression parsing, we might want to wait until https://github.com/typetools/checker-framework/pull/2062 is merged.

--- a/checker/src/main/java/org/checkerframework/checker/index/inequality/LessThanTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/inequality/LessThanTransfer.java
@@ -60,7 +60,7 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 // right is already bottom, nothing to refine.
                 return;
             }
-            if (isDoubleOrFloatLiteral(leftRec)) {
+            if (!isDoubleOrFloatLiteral(leftRec)) {
                 lessThanExpressions.add(leftRec.toString());
             }
             Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
@@ -92,7 +92,7 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 // right is already bottom, nothing to refine.
                 return;
             }
-            if (isDoubleOrFloatLiteral(leftRec)) {
+            if (!isDoubleOrFloatLiteral(leftRec)) {
                 lessThanExpressions.add(leftRec.toString() + " + 1");
             }
             Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
@@ -116,7 +116,7 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 if (expressions == null) {
                     expressions = new ArrayList<>();
                 }
-                if (isDoubleOrFloatLiteral(leftRec)) {
+                if (!isDoubleOrFloatLiteral(leftRec)) {
                     expressions.add(leftRec.toString());
                 }
                 AnnotationMirror refine = factory.createLessThanQualifier(expressions);
@@ -147,10 +147,10 @@ public class LessThanTransfer extends IndexAbstractTransfer {
      */
     private boolean isDoubleOrFloatLiteral(Receiver receiver) {
         if (receiver instanceof ValueLiteral) {
-            return receiver.getType().getKind() != TypeKind.DOUBLE
-                    && receiver.getType().getKind() != TypeKind.FLOAT;
+            return receiver.getType().getKind() == TypeKind.DOUBLE
+                    || receiver.getType().getKind() == TypeKind.FLOAT;
         } else {
-            return true;
+            return false;
         }
     }
 }

--- a/checker/src/main/java/org/checkerframework/checker/index/inequality/LessThanTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/inequality/LessThanTransfer.java
@@ -5,11 +5,13 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Set;
 import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.type.TypeKind;
 import org.checkerframework.checker.index.IndexAbstractTransfer;
 import org.checkerframework.checker.index.IndexUtil;
 import org.checkerframework.common.value.ValueAnnotatedTypeFactory;
 import org.checkerframework.dataflow.analysis.FlowExpressions;
 import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
+import org.checkerframework.dataflow.analysis.FlowExpressions.ValueLiteral;
 import org.checkerframework.dataflow.analysis.RegularTransferResult;
 import org.checkerframework.dataflow.analysis.TransferInput;
 import org.checkerframework.dataflow.analysis.TransferResult;
@@ -18,6 +20,7 @@ import org.checkerframework.dataflow.cfg.node.NumericalSubtractionNode;
 import org.checkerframework.framework.flow.CFAnalysis;
 import org.checkerframework.framework.flow.CFStore;
 import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.util.FlowExpressionParseUtil;
 
 /**
  * Implements 3 refinement rules:
@@ -57,9 +60,23 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 // right is already bottom, nothing to refine.
                 return;
             }
-            lessThanExpressions.add(leftRec.toString());
+            if (isParsable(leftRec)) {
+                lessThanExpressions.add(leftRec.toString());
+            }
             Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
             store.insertValue(rightRec, factory.createLessThanQualifier(lessThanExpressions));
+        }
+    }
+
+    /**
+     * Return true if {@code receiver}'s toString can be parsed by {@link FlowExpressionParseUtil}.
+     */
+    private boolean isParsable(Receiver receiver) {
+        if (receiver instanceof ValueLiteral) {
+            return receiver.getType().getKind() != TypeKind.DOUBLE
+                    && receiver.getType().getKind() != TypeKind.FLOAT;
+        } else {
+            return true;
         }
     }
 
@@ -87,7 +104,9 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 // right is already bottom, nothing to refine.
                 return;
             }
-            lessThanExpressions.add(leftRec.toString() + " + 1");
+            if (isParsable(leftRec)) {
+                lessThanExpressions.add(leftRec.toString() + " + 1");
+            }
             Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
             store.insertValue(rightRec, factory.createLessThanQualifier(lessThanExpressions));
         }
@@ -109,7 +128,9 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 if (expressions == null) {
                     expressions = new ArrayList<>();
                 }
-                expressions.add(leftRec.toString());
+                if (isParsable(leftRec)) {
+                    expressions.add(leftRec.toString());
+                }
                 AnnotationMirror refine = factory.createLessThanQualifier(expressions);
                 CFValue value = analysis.createSingleAnnotationValue(refine, n.getType());
                 CFStore info = in.getRegularStore();

--- a/checker/src/main/java/org/checkerframework/checker/index/inequality/LessThanTransfer.java
+++ b/checker/src/main/java/org/checkerframework/checker/index/inequality/LessThanTransfer.java
@@ -60,23 +60,11 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 // right is already bottom, nothing to refine.
                 return;
             }
-            if (isParsable(leftRec)) {
+            if (isDoubleOrFloatLiteral(leftRec)) {
                 lessThanExpressions.add(leftRec.toString());
             }
             Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
             store.insertValue(rightRec, factory.createLessThanQualifier(lessThanExpressions));
-        }
-    }
-
-    /**
-     * Return true if {@code receiver}'s toString can be parsed by {@link FlowExpressionParseUtil}.
-     */
-    private boolean isParsable(Receiver receiver) {
-        if (receiver instanceof ValueLiteral) {
-            return receiver.getType().getKind() != TypeKind.DOUBLE
-                    && receiver.getType().getKind() != TypeKind.FLOAT;
-        } else {
-            return true;
         }
     }
 
@@ -104,7 +92,7 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 // right is already bottom, nothing to refine.
                 return;
             }
-            if (isParsable(leftRec)) {
+            if (isDoubleOrFloatLiteral(leftRec)) {
                 lessThanExpressions.add(leftRec.toString() + " + 1");
             }
             Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
@@ -128,7 +116,7 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                 if (expressions == null) {
                     expressions = new ArrayList<>();
                 }
-                if (isParsable(leftRec)) {
+                if (isDoubleOrFloatLiteral(leftRec)) {
                     expressions.add(leftRec.toString());
                 }
                 AnnotationMirror refine = factory.createLessThanQualifier(expressions);
@@ -150,6 +138,19 @@ public class LessThanTransfer extends IndexAbstractTransfer {
                     factory.getQualifierHierarchy().findAnnotationInHierarchy(s, factory.UNKNOWN));
         } else {
             return Collections.emptyList();
+        }
+    }
+
+    /**
+     * Return true if {@code receiver} is a double or float literal, which can't be parsed by {@link
+     * FlowExpressionParseUtil}.
+     */
+    private boolean isDoubleOrFloatLiteral(Receiver receiver) {
+        if (receiver instanceof ValueLiteral) {
+            return receiver.getType().getKind() != TypeKind.DOUBLE
+                    && receiver.getType().getKind() != TypeKind.FLOAT;
+        } else {
+            return true;
         }
     }
 }

--- a/checker/tests/index/Issue2030.java
+++ b/checker/tests/index/Issue2030.java
@@ -1,0 +1,9 @@
+public class Issue2030 {
+    double roundIntermediate(double x) {
+        if (x >= 0.0) {
+            return x;
+        } else {
+            return (long) x - 1;
+        }
+    }
+}

